File: Assets/Standard Assets/Logic/Museum/Museum.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System;
namespace KBN
{
	public abstract class Museum
	{
		protected HashObject artifaceData;
		protected List<EventEntity> artifactArray;
		
		protected string noEventDes;
		protected List<EventEntity> eventsByOrder;
		protected int hasNewEvent=0;
		public static Museum singleton { get; protected set; }


		public List<EventEntity> getOrderedEvents
		{
			get 
			{
				return eventsByOrder;
			}
		}
		
		public List<EventEntity> getArtiFacts
		{
			get 
			{
				return artifactArray;
			}
		}
		public void ReSetArtDatas(int id,int count){
			if(artifactArray!=null){
				for(int i=0;i<artifactArray.Count;i++){
					if(id==artifactArray[i].id){
						artifactArray[i].resetArtEvent(count);
					}else{
						artifactArray[i].resetEvent();
					}
				}
			}
		}
		public EventEntity GetEventById(int id){
			if(eventsByOrder!=null){
				for(int i=0;i<eventsByOrder.Count;i++){
					if(id==eventsByOrder[i].id){
						return eventsByOrder[i];
					}
				}
			}
			return null;
		}
		public EventEntity GetArtById(int id){
			if(artifactArray!=null){
				for(int i=0;i<artifactArray.Count;i++){
					if(id==artifactArray[i].id){
						return artifactArray[i];
					}
				}
			}
			return null;
		}
		public void checkForRoundTower()
		{
			
		}

		protected int count = 0;
		
		public bool HasEvent
		{
			get
			{
				return hasNewEvent > 0 ? true : false;
			}
		}
		
		public string getNoEventDes
		{
			get 
			{
				return noEventDes;
			}
		}
		
		public int eventNum
		{
			get
			{
				return hasNewEvent;
			}
		}
		
		public bool isItemExist(int itemId)
		{
			
			if(itemId >= 20410 && itemId <= 21000)
			{
				return true;
			}
			else if( itemId >= 21100 && itemId <= 21103 ) // World map bonus items
			{
				return true;
			}
			
			return false;
		}

		public void OnMutiClaimOk(HashObject result)
		{
			if(_Global.GetBoolean(result["ok"]))
			{
				int eventId = _Global.INT32(result["eventId"]);
				int crestId = _Global.INT32(result["crestId"]);
				int claimCount = _Global.INT32(result["claimCount"]);
				EventEntity eventEntity;
				for(int a = 0; a < eventsByOrder.Count; a ++)
				{
					eventEntity = eventsByOrder[a];
					
					if(eventEntity.id == eventId)
					{
						MyItems.singleton.AddItem(eventEntity.itemId, eventEntity.itemNum*claimCount);
						
						EventEntity.EventPiece piece;
						for(int b = 0; b < eventEntity.pieces.Count; b++)
						{
							piece = eventEntity.pieces[b];
							if (crestId > 0) {				// 'or' logic
								if (crestId == piece.id) {
									MyItems.singleton.subtractItem(piece.id, (int)piece.needNum*claimCount);
									break;
								}
							} else {						// 'and' logic
								MyItems.singleton.subtractItem(piece.id, (int)piece.needNum*claimCount);
							}
						}
						
						break;
					}
				}
				
				for(int a = 0; a < eventsByOrder.Count; a ++)
				{
					eventEntity = eventsByOrder[a];
					
					eventEntity.resetEvent();
					
					if(eventEntity.id == eventId)
					{
						eventEntity.resetChangeQuanlity(claimCount);
					}
				}
				
//				if(callback != null)
//				{
//					callback.DynamicInvoke();
//				}
			}
		}

		public void MutiClaimEvent(int eventId, int crestId, int claimNum, MulticastDelegate okFunc)
		{
			if (crestId > 0)
				UnityNet.reqClaimEvent(eventId, crestId,claimNum, okFunc, null);	
			else
				UnityNet.reqClaimEvent(eventId,claimNum, okFunc, null);	
		}

		protected HashObject InitArtifactData()
		{
			if(artifaceData == null)
			{
				artifaceData = Datas.singleton.artifactData();
			}
			return artifaceData;
		}
		public string loadArtifactTexture(int category)
		{
			return TextureMgr.instance().LoadTileNameFromSetting("" + category);
		}
		public TileSprite loadArtifactSprite(int category)
		{
			return TextureMgr.instance().GetSpriteFromSetting(""+ category);
		}
		
		public int ComputeClaimCount(EventEntity eventEntity)
		{
			int iCount = int.MaxValue;
			int exchangeCount = 0;
			EventEntity.EventPiece piece;
			for(int i=0;i<eventEntity.pieces.Count;i++)
			{
				piece = eventEntity.pieces[i] as EventEntity.EventPiece;
				exchangeCount = (int)(piece.ownNum/piece.needNum);
				iCount = exchangeCount<iCount?exchangeCount:iCount;
			}
			return iCount;
		}	
	}
	public class Artifact
	{
		public List<Piece> pieces;
		public int category;
		public string icon;
		public string tips;
		public string des;
		public int price;
		public TileSprite sprite;
		public int artifactId;

		public int priority;
		
		public bool initArtifact(HashObject data)
		{
			category = _Global.INT32(data["rewardItemId"].Value);
			
			if(category == 0)
			{
				return false;
			}
			
			artifactId = _Global.INT32(data["eventId"].Value);
			icon	 = TextureMgr.instance().LoadTileNameOfItem(category);
			sprite	 = Museum.singleton.loadArtifactSprite(category);
			tips	 = Datas.getArString("itemName.i" + category);
			des		 = Datas.getArString("itemDesc.i" + category);
			priority = _Global.INT32(data["priority"].Value);

			pieces = new List<Piece>();
			
			object[] arr = _Global.GetObjectValues(data["exchangings"]);	
			Piece piece;
			HashObject pieceObj;
			for(int a = 0; a < arr.Length; a ++)
			{
				pieceObj = data["exchangings"][_Global.ap + a];
				piece = new Piece();
				piece.id = _Global.INT32(pieceObj["crestId"].Value);
				
				if(piece.id == 0)
				{
					continue;
				}			
				
				piece.need = _Global.INT32(pieceObj["amount"].Value);
				piece.own = MyItems.singleton.countForItem(piece.id);
				piece.icon = TextureMgr.instance().LoadTileNameOfItem(piece.id);
				piece.name = Datas.getArString("itemName.i" + piece.id);

				pieces.Add(piece);
			}

			return true;		
		}
		
		public void resetPieces()
		{
			if(pieces == null)
			{
				pieces = new List<Piece>();
			}
			
			Piece piece;
			for(int i = 0; i < pieces.Count; i++)
			{
				piece = pieces[i] as Piece;
				piece.own = MyItems.singleton.countForItem(piece.id);
			}		
		}
		
		public bool CanClaim()
		{
			bool result = true;
			for(int i=0;i< pieces.Count;i++)
			{	
				if((pieces[i] as Piece).own < (pieces[i] as Piece).need)
				{
					result = false;
					break;
				}
			}
			if(result)
			{
				List<ArtifactClaimRequireItem> requires = ClaimRequireMents();
				for(int j=0;j<requires.Count;j++)
				{
					if((requires[j] as ArtifactClaimRequireItem).completed == false)
					{
						result = false;
						break;
					}
				}
			}
			return result;
		}
		//TODO: fix next version
		public List<ArtifactClaimRequireItem> ClaimRequireMents()
		{
			ArtifactClaimRequireItem requireResearch = new ArtifactClaimRequireItem();
			requireResearch.completed = true;
			requireResearch.require = "";
			ArtifactClaimRequireItem requireBuilding = new ArtifactClaimRequireItem();
			requireBuilding.completed = true;
			requireBuilding.require = "";
			List<ArtifactClaimRequireItem> result = new List<ArtifactClaimRequireItem>();
			result.Add(requireResearch);
			result.Add(requireBuilding);
			return result;
		}
		
	}
	public class Piece
	{
		public long need;
		public long own;
		public int id;
		public string name;
		public string icon;
		public int price;
	}
	public class ArtifactClaimRequireItem
	{
		public string require;
		public bool completed;
	}
	
	public class EventEntity
	{
		public class EventPiece:UnityEngine.Object
		{
			public int id;
			public int categary;
			public string texturePath;
			public long ownNum;
			public long needNum;
			public bool IsOrLogic;
		}
		
		public int id;
		public int tab;
		
		public string eventName;
		public int limitQuantity;
		public int changedQuantity;
		
		public long startTime;
		public long endTime;
		public long needTime;
		
		public int itemId;
		public int itemCategary;
		public string itemTexture;
		public int itemNum;
		public int logicType; // Constant.Museum.singleton.ExchangeLogicType
		public bool batchReward;
		public List<EventPiece> pieces = new List<EventPiece>();
		
		public bool canClaim = false;

		public int priority;

		public void resetPieces()
		{
			if(pieces == null)
			{
				pieces = new List<EventPiece>();
			}
			
			EventPiece piece;
			for(int i = 0; i < pieces.Count; i++)
			{
				piece = pieces[i] as EventPiece;
				piece.ownNum = MyItems.singleton.countForItem(piece.id);
			}		
		}

		public bool CanClaim()
		{
			bool result = true;
			for(int i=0;i< pieces.Count;i++)
			{	
				if((pieces[i] as EventPiece).ownNum < (pieces[i] as EventPiece).needNum)
				{
					result = false;
					break;
				}
			}
			if(result)
			{
				List<ArtifactClaimRequireItem> requires = ClaimRequireMents();
				for(int j=0;j<requires.Count;j++)
				{
					if((requires[j] as ArtifactClaimRequireItem).completed == false)
					{
						result = false;
						break;
					}
				}
			}
			return result;
		}
		public List<ArtifactClaimRequireItem> ClaimRequireMents()
		{
			ArtifactClaimRequireItem requireResearch = new ArtifactClaimRequireItem();
			requireResearch.completed = true;
			requireResearch.require = "";
			ArtifactClaimRequireItem requireBuilding = new ArtifactClaimRequireItem();
			requireBuilding.completed = true;
			requireBuilding.require = "";
			List<ArtifactClaimRequireItem> result = new List<ArtifactClaimRequireItem>();
			result.Add(requireResearch);
			result.Add(requireBuilding);
			return result;
		}

		public bool initEvent(HashObject param)
		{
			HashObject data = param as HashObject;
			
			id = _Global.INT32(data["eventId"].Value);
			startTime = data["startUnixtime"]==null?0:_Global.INT64(data["startUnixtime"].Value);
			endTime = data["endUnixtime"]==null?0:(_Global.INT64(data["endUnixtime"].Value) - 60);
			batchReward = data["batchReward"]==null?false:_Global.INT32(data["batchReward"])==1;
			priority = _Global.INT32(data["priority"].Value);

			if (data["tab"] != null)
			{
				tab = _Global.INT32(data["tab"]);
			}
			else
			{
				tab = 1;
			}
			
			if(data["rewardItemId"].Value == null)
			{
				return false;
			}
			
			itemId = _Global.INT32(data["rewardItemId"].Value);
			itemNum = _Global.INT32(data["rewardItemAmount"].Value); 
			itemCategary = getCategary(itemId);
			itemTexture	 = TextureMgr.instance().LoadTileNameOfItem(itemId);//getTexturePath(itemCategary, itemId);


			if(data["eventName"] != null)
			{
				eventName = data["eventName"].Value + "";
			}
			
			if(data["rewardLimit"] != null)
			{
				limitQuantity = _Global.INT32(data["rewardLimit"].Value);
			}
			
			if(data["rewardCount"] != null)
			{
				changedQuantity = _Global.INT32(data["rewardCount"].Value);
			}
			
			logicType = (data["type"] != null ? _Global.INT32(data["type"]) : /*Constant.Museum.singleton.ExchangeLogicType.LOGIC_OR);*/ Constant.Museum.ExchangeLogicType.LOGIC_AND);
			
			var isAndLogic = (logicType == Constant.Museum.ExchangeLogicType.LOGIC_AND);
			canClaim = isAndLogic;
			
			object[] arr = _Global.GetObjectValues(data["exchangings"]);		
			EventPiece piece;
			HashObject pieceObj;
			for(int a = 0; a < arr.Length; a++)
			{
				pieceObj = data["exchangings"][_Global.ap + a];
				piece = new EventPiece();
				piece.id = _Global.INT32(pieceObj["crestId"].Value);
				piece.needNum = _Global.INT32(pieceObj["amount"].Value);
				piece.ownNum  = MyItems.singleton.countForItem(piece.id); 
				piece.categary = getCategary(piece.id);
				piece.texturePath = TextureMgr.instance().LoadTileNameOfItem(piece.id);	
				piece.IsOrLogic=logicType == Constant.Museum.ExchangeLogicType.LOGIC_OR;

				pieces.Add(piece);

				if(isAndLogic && piece.needNum > piece.ownNum)
				{
					canClaim = false;

				} 
				else if(!isAndLogic && piece.needNum <= piece.ownNum)
				{
					canClaim = true;
				}
			}
			if(limitQuantity!=0){
				if(changedQuantity >= limitQuantity)
				{
					canClaim = false;
				}
			}
			
			return true;
		}
		public void resetArtEvent(int count)
		{
			EventPiece piece;
			
			bool isAndLogic = (logicType == Constant.Museum.ExchangeLogicType.LOGIC_AND);
			canClaim = isAndLogic;
			
			for(int a = 0; a < pieces.Count; a ++)
			{
				piece = pieces[a];
				
				piece.ownNum = MyItems.singleton.countForItem(piece.id);
				
				if(isAndLogic && piece.needNum > piece.ownNum)
				{
					canClaim = false;
				}
				else if(!isAndLogic && piece.needNum <= piece.ownNum)
				{
					canClaim = true;
				}
			}
			
			if(limitQuantity!=0){
				if(changedQuantity >= limitQuantity)
				{
					canClaim = false;
				}
			}
			resetChangeQuanlity(count);		
		}
		
		public void resetEvent()
		{
			EventPiece piece;
			
			bool isAndLogic = (logicType == Constant.Museum.ExchangeLogicType.LOGIC_AND);
			canClaim = isAndLogic;
			
			for(int a = 0; a < pieces.Count; a ++)
			{
				piece = pieces[a];
				
				piece.ownNum = MyItems.singleton.countForItem(piece.id);
				
				if(isAndLogic && piece.needNum > piece.ownNum)
				{
					canClaim = false;
				}
				else if(!isAndLogic && piece.needNum <= piece.ownNum)
				{
					canClaim = true;
				}
			}
			
			if(limitQuantity!=0){
				if(changedQuantity >= limitQuantity)
				{
					canClaim = false;
				}
			}		
		}
		
		public void resetChangeQuanlity(int changeCount)
		{		
			changedQuantity += changeCount;
			
			if(limitQuantity!=0){
				if(changedQuantity >= limitQuantity)
				{
					canClaim = false;
				}
			}	
		}
		
		public int getCategary(int _itemId)
		{
			HashObject item = (Datas.singleton.itemlist())["i" + _itemId];
			
			// Hot fix for v16.3
			if (item == null) {
				return (int)MyItems.Category.UnknowItem;
			}
			
			int returnCategory;
			
			if(MystryChest.singleton.IsMystryChest(_itemId))
			{
				returnCategory = (int)MyItems.Category.MystryChest;
			}
			else if(MystryChest.singleton.IsLevelChest(_itemId))
			{
				returnCategory = (int)MyItems.Category.LevelChest;
			}
			else
			{
				returnCategory = _Global.INT32(item["category"]);
			}
			
			return returnCategory;			
		}
		
		public string getTexturePath(int _categary,int _id)
		{
			string imageName = "";
			
			if(_categary == (int)MyItems.Category.MystryChest)
			{ 
				imageName = MystryChest.singleton.GetChestImage(_id);
				if(!TextureMgr.instance().ItemSpt().IsTileExist(imageName))
				{
					imageName = Constant.DefaultChestTileName;
				}			
			}
			else if(_categary == (int)MyItems.Category.LevelChest)
			{
				imageName = MystryChest.singleton.GetLevelChestImage(_id);
				if(!TextureMgr.instance().ItemSpt().IsTileExist(imageName))
				{
					imageName = Constant.DefaultChestTileName;
				}			
			}
			else
			{
				if(_categary == (int)MyItems.Category.Chest)
				{
					imageName = "i"+ _id;
					if(!TextureMgr.instance().ItemSpt().IsTileExist(imageName))
					{
						imageName = Constant.DefaultChestTileName;
					}		
				}
				else
				{
					if(Datas.singleton.getImageName(_id) == "")
					{
						imageName = "i"+ _id;
					}
					else
					{
						imageName =	Datas.singleton.getImageName(_id);
					}
				}
			}
			
			return imageName;				
		}
	}
}

File: Assets/Script/Map/Museum.js
import System.Collections;
import System.Collections.Generic;
import System.Reflection;
public class Museum extends KBN.Museum
{
	public	static	function	instance()
	{
		if( singleton == null ){
			singleton = new Museum();
		}
		return singleton as Museum;
	}

	public function init(data:HashObject):void
	{	
		noEventDes = Datas.getArString("Museum.NoCurrentEvents");
	
		if(data["rtEvent"] != null)
		{
			hasNewEvent = _Global.INT32(data["rtEvent"].Value);
			
			if(hasNewEvent > 0)
			{
				var decInfor:BuildingDecMgr.DecrationInfor = new BuildingDecMgr.DecrationInfor();
				decInfor.buildingType = Constant.Building.MUSEUM;
				decInfor.textureName = "RoundTower_icon2";
				decInfor.sceneLevel = GameMain.CITY_SCENCE_LEVEL;
				decInfor.slotId = Building.instance().getPositionForType(Constant.Building.MUSEUM);
				ChromeOrganizerMenu.IsShowRoundTowerButton = true;
				
				BuildingDecMgr.getInstance().addDecration(decInfor);					
			}
			else
			{
				ChromeOrganizerMenu.IsShowRoundTowerButton = false;
			}		
		}
		else
		{
			hasNewEvent = 0;
		}
																	
	}
	
	public function updateEvent(data:HashObject):void
	{
	/*
		var a:int = _Global.INT32(data.Value);
	
		if(count++ == 0)
		{
			a = 0;
			count = 0;
		}
	*/
		//if(hasNewEvent != a)
		if(hasNewEvent != _Global.INT32(data.Value))
		{			
			hasNewEvent = _Global.INT32(data.Value);
			
			if(hasNewEvent > 0)
			{
				var decInfor:BuildingDecMgr.DecrationInfor = new BuildingDecMgr.DecrationInfor();
				decInfor.buildingType = Constant.Building.MUSEUM;
				decInfor.textureName = "RoundTower_icon2";
				decInfor.sceneLevel = GameMain.CITY_SCENCE_LEVEL;
				decInfor.slotId = Building.instance().getPositionForType(Constant.Building.MUSEUM);
				
				BuildingDecMgr.getInstance().addDecration(decInfor);
				ChromeOrganizerMenu.IsShowRoundTowerButton = true;
				
			}
			else
			{
				BuildingDecMgr.getInstance().delDecration(Constant.Building.MUSEUM);
				ChromeOrganizerMenu.IsShowRoundTowerButton = false;
			}
		}
	}
	
	public function ClaimEvent(eventId:int, crestId:int, claimNum:int, callback:System.MulticastDelegate):void
	{	
		var ok:Function = function(result:HashObject)
		{
			if(result["ok"].Value)
			{
				var eventEntity:KBN.EventEntity;
				for(var a:int = 0; a < eventsByOrder.Count; a ++)
				{
					eventEntity = eventsByOrder[a];

					if(eventEntity.id == eventId)
					{
						MyItems.instance().AddItem(eventEntity.itemId, eventEntity.itemNum*claimNum);
						
						var piece:KBN.EventEntity.EventPiece;
						for(var b:int = 0; b < eventEntity.pieces.Count; b++)
						{
							piece = eventEntity.pieces[b];
							if (crestId > 0) {				// 'or' logic
								if (crestId == piece.id) {
									MyItems.instance().subtractItem(piece.id, piece.needNum*claimNum);
									break;
								}
							} else {						// 'and' logic
								MyItems.instance().subtractItem(piece.id, piece.needNum*claimNum);
							}
						}

						break;
					}
				}
				
				for(a = 0; a < eventsByOrder.Count; a ++)
				{
					eventEntity = eventsByOrder[a];
				
					eventEntity.resetEvent();
					
					if(eventEntity.id == eventId)
					{
						eventEntity.resetChangeQuanlity(claimNum);
					}
				}
				
				if(callback != null)
				{
					callback.DynamicInvoke();
				}
			}
		};
		if (crestId > 0)
			UnityNet.reqClaimEvent(eventId, crestId,claimNum, ok, null);	
		else
			UnityNet.reqClaimEvent(eventId,claimNum, ok, null);	
	}
	
	public function GetEvents(callback:Function):void
	{
		var ok:Function = function(result:HashObject){
			if(result["ok"].Value)
			{
				var exchanges:HashObject = result["data"]["events"];
				var events:Array = _Global.GetObjectValues(exchanges);
				
				eventsByOrder = new List.<KBN.EventEntity>();
				
				var hashObj:HashObject = new HashObject({"num":events.length});
				updateEvent(hashObj["num"] as HashObject);
								
				if(events.length == 0)
				{
					noEventDes = result["data"]["note"].Value as String;
					
					if(noEventDes == "")					
					{
						noEventDes = Datas.getArString("Museum.NoCurrentEvents");
					}
				}
				else
				{
					var eventObj:HashObject;
					var eventDetail:KBN.EventEntity;
					var index:int = 0;
					for(var i=0; i<events.length; i++)
					{
						eventObj = exchanges[_Global.ap + i] as HashObject;

						eventDetail = new KBN.EventEntity();
						if(!eventDetail.initEvent(eventObj))
						{
							continue;
						}
						
						eventsByOrder.Add(eventDetail);												
					}
					EventPriority();
				}

				if(callback)
				{
					callback(eventsByOrder);
				}
			}
		};
		
		UnityNet.reqEventDetail(ok, null);		
	}
	public function EventPriority():void
	{
		var higherPriorityList = new List.<KBN.EventEntity>();
		var lowerPriorityList = new List.<KBN.EventEntity>();
		for(var j=0; j<eventsByOrder.Count;j++)
		{
			if(eventsByOrder[j].priority == 0)
			{
				higherPriorityList.Add(eventsByOrder[j]);
			}
			else
			{
				lowerPriorityList.Add(eventsByOrder[j]);
			}
		}
		higherPriorityList = EventOrder(higherPriorityList);
		lowerPriorityList = EventOrder(lowerPriorityList);
		for(var m=0;m<lowerPriorityList.Count;m++)
		{
			higherPriorityList.Add(lowerPriorityList[m]);
		}
		eventsByOrder = higherPriorityList;
	}
	public function EventOrder(priorityArray:List.<KBN.EventEntity>):List.<KBN.EventEntity>
	{
		var canClainList = new List.<KBN.EventEntity>();
		var canNotClaimList = new List.<KBN.EventEntity>();
		for(var j=0; j<priorityArray.Count;j++)
		{
			if(priorityArray[j].canClaim == false)
			{
				canNotClaimList.Add(priorityArray[j]);
			}
			else
			{
				canClainList.Add(priorityArray[j]);
			}
		}
		for(var m=0;m<canNotClaimList.Count;m++)
		{
			canClainList.Add(canNotClaimList[m]);
		}
		priorityArray = canClainList;
		return priorityArray;
	}
	
	public function ArifactPriority():void
	{
		var higherPriorityList = new List.<KBN.EventEntity>();
		var lowerPriorityList = new List.<KBN.EventEntity>();
		for(var j=0; j<artifactArray.Count;j++)
		{
			if(artifactArray[j].priority == 0)
			{
				higherPriorityList.Add(artifactArray[j]);
			}
			else
			{
				lowerPriorityList.Add(artifactArray[j]);
			}
		}
		higherPriorityList = ArtifactOrder(higherPriorityList);
		lowerPriorityList = ArtifactOrder(lowerPriorityList);
		for(var m=0;m<lowerPriorityList.Count;m++)
		{
			higherPriorityList.Add(lowerPriorityList[m]);
		}
		artifactArray = higherPriorityList;
	}
	public function ArtifactOrder(priorityArray:List.<KBN.EventEntity>):List.<KBN.EventEntity>
	{
		var canClainArt = new List.<KBN.EventEntity>();
		var canNotClaimArt = new List.<KBN.EventEntity>();
		for(var i=0; i<priorityArray.Count;i++)
		{
			if(priorityArray[i].CanClaim() == false)
			{
				canNotClaimArt.Add(priorityArray[i]);
			}
			else
			{
				canClainArt.Add(priorityArray[i]);
			}
		}
		for(var j=0;j<canNotClaimArt.Count;j++)
		{
			canClainArt.Add(canNotClaimArt[j]);
		}
		priorityArray = canClainArt;
		return priorityArray;	
	}
	public function GetArtifacts(callback:Function):void
	{
		
		var ok:Function = function(result:HashObject){
			if(result["ok"].Value)
			{
				var exchanges:HashObject = result["data"]["events"];
				var artifacts:Array = _Global.GetObjectValues(exchanges);
				
				artifactArray = new List.<KBN.EventEntity>();
				
				var artifactObj:HashObject;
				var artifact:KBN.EventEntity;
				var index:int = 0;
				for(var i=0; i<artifacts.length; i++)
				{
					artifactObj = exchanges[_Global.ap + i] as HashObject;

					artifact = new KBN.EventEntity();
					
					if(!artifact.initEvent(artifactObj))
					{
						continue;
					}
					
					artifactArray.Add(artifact);
				}
				ArifactPriority();

				if(callback)
				{
					callback(artifactArray);
				}
			}
		};	
		
		UnityNet.reqArtifactDetail(ok, null);		

	/*
		if(artifactArray == null || artifactArray.length == 0)
		{
			artifactArray = new Array();
			piecesShow = new Array();
			var artifactDataInternal:HashObject = InitArtifactData();
			var showCategories:Array = _Global.GetObjectValues(artifactDataInternal["showCategories"]);
			var piecesData:HashObject = artifactDataInternal["artifactNeedCount"];
			var artifactCategories:HashObject = artifactDataInternal["artifactCategories"];
			
			for(var i=0;i<showCategories.length;i++)
			{
				var artifactIds:Array = _Global.GetObjectValues(artifactCategories["" + (showCategories[i] as HashObject).Value]);
	
				for(var j=0;j<artifactIds.length;j++)
				{
					var artifactId:int =_Global.INT32((artifactIds[j] as HashObject));
					var artifact:Artifact = new Artifact();
					artifact.category = artifactId;
					artifact.tips = Datas.getArString("artifactPieceName.a" + artifactId);
					artifact.icon = loadArtifactTexture(artifactId);
					artifact.des = Datas.getArString("artifactSetDesc.a" + artifactId);
					artifact.sprite = loadArtifactSprite(artifactId);
					//pieces
					artifact.pieces = new Array();
					
					var artifactPieces:Array = _Global.GetObjectValues(piecesData["" + artifactId]);
					for(var k=0;k<artifactPieces.length;k++)
					{
						var piece:Piece = new Piece();
						var pieceIdAndCount:Array = _Global.GetObjectValues((artifactPieces[k] as HashObject));
						piece.id = _Global.INT32((pieceIdAndCount[0] as HashObject));
						piece.need = _Global.INT32((pieceIdAndCount[1] as HashObject));
						piece.name = Datas.getArString("artifactPieceName.a" + piece.id);
						piece.price = Shop.instance.getPriceOfItem(Shop.PIECE,piece.id, true);
						piece.icon = "i" + piece.id;
						piece.own = MyItems.instance().countForItem(piece.id);
						artifact.pieces.Add(piece);
						piecesShow.Add(piece.id);
					}
					artifactArray.Add(artifact);
				}
			}
			
			return artifactArray;
		
		}
		else
		{
			for(var l =0;l< artifactArray.length;l++)
			{
				var arcifact:Artifact = (artifactArray[l] as Artifact);
				for(var m = 0;m<arcifact.pieces.length;m++)
				{
					(arcifact.pieces[m] as Piece).own = MyItems.instance().countForItem( (arcifact.pieces[m] as Piece).id);
				}
			}
				
			return artifactArray;
		}*/
		
	}
	
				
}